File: test/interface/golden/sum/SumWithRecord.ml
type sumWithRecordA1 =
  { a1 : int
  }

type sumWithRecordB2 =
  { b2 : string
  ; b3 : int
  }

type sumWithRecord =
  | A1 of sumWithRecordA1
  | B2 of sumWithRecordB2

let encodeSumWithRecordA1 x =
  Aeson.Encode.object_
    [ ( "a1", Aeson.Encode.int x.a1 )
    ]

let encodeSumWithRecordB2 x =
  Aeson.Encode.object_
    [ ( "b2", Aeson.Encode.string x.b2 )
    ; ( "b3", Aeson.Encode.int x.b3 )
    ]

let encodeSumWithRecord x =
  match x with
  | A1 y0 ->
     (match (Js.Json.decodeObject (encodeSumWithRecordA1 y0)) with
      | Some dict ->
         Js.Dict.set dict "tag" (Js.Json.string "A1");
         Js.Json.object_ dict
      | None ->
         Aeson.Encode.object_ []
     )
  | B2 y0 ->
     (match (Js.Json.decodeObject (encodeSumWithRecordB2 y0)) with
      | Some dict ->
         Js.Dict.set dict "tag" (Js.Json.string "B2");
         Js.Json.object_ dict
      | None ->
         Aeson.Encode.object_ []
     )

let decodeSumWithRecordA1 json =
  match Aeson.Decode.
    { a1 = field "a1" int json
    }
  with
  | v -> Js_result.Ok v
  | exception Aeson.Decode.DecodeError message -> Js_result.Error ("decodeSumWithRecordA1: " ^ message)

let decodeSumWithRecordB2 json =
  match Aeson.Decode.
    { b2 = field "b2" string json
    ; b3 = field "b3" int json
    }
  with
  | v -> Js_result.Ok v
  | exception Aeson.Decode.DecodeError message -> Js_result.Error ("decodeSumWithRecordB2: " ^ message)

let decodeSumWithRecord json =
  match Aeson.Decode.(field "tag" string json) with
  | "A1" ->
     (match decodeSumWithRecordA1 json with
      | Js_result.Ok v -> Js_result.Ok (A1 v)
      | Js_result.Error message -> Js_result.Error ("decodeSumWithRecord: " ^ message)
     )
  | "B2" ->
     (match decodeSumWithRecordB2 json with
      | Js_result.Ok v -> Js_result.Ok (B2 v)
      | Js_result.Error message -> Js_result.Error ("decodeSumWithRecord: " ^ message)
     )
  | err -> Js_result.Error ("Unknown tag value found '" ^ err ^ "'.")
  | exception Aeson.Decode.DecodeError message -> Js_result.Error message

File: test/interface/golden/lib/js/sum/SumWithRecord.js
'use strict';

var Block        = require("bs-platform/lib/js/block.js");
var Js_exn       = require("bs-platform/lib/js/js_exn.js");
var Js_json      = require("bs-platform/lib/js/js_json.js");
var Aeson_decode = require("bs-aeson/lib/js/src/Aeson_decode.js");
var Aeson_encode = require("bs-aeson/lib/js/src/Aeson_encode.js");

function encodeSumWithRecordA1(x) {
  return Aeson_encode.object_(/* :: */[
              /* tuple */[
                "a1",
                x[/* a1 */0]
              ],
              /* [] */0
            ]);
}

function encodeSumWithRecordB2(x) {
  return Aeson_encode.object_(/* :: */[
              /* tuple */[
                "b2",
                x[/* b2 */0]
              ],
              /* :: */[
                /* tuple */[
                  "b3",
                  x[/* b3 */1]
                ],
                /* [] */0
              ]
            ]);
}

function encodeSumWithRecord(x) {
  if (x.tag) {
    var match = Js_json.decodeObject(encodeSumWithRecordB2(x[0]));
    if (match) {
      var dict = match[0];
      dict["tag"] = "B2";
      return dict;
    } else {
      return Aeson_encode.object_(/* [] */0);
    }
  } else {
    var match$1 = Js_json.decodeObject(encodeSumWithRecordA1(x[0]));
    if (match$1) {
      var dict$1 = match$1[0];
      dict$1["tag"] = "A1";
      return dict$1;
    } else {
      return Aeson_encode.object_(/* [] */0);
    }
  }
}

function decodeSumWithRecordA1(json) {
  var exit = 0;
  var v;
  try {
    v = /* record */[/* a1 */Aeson_decode.field("a1", Aeson_decode.$$int, json)];
    exit = 1;
  }
  catch (raw_exn){
    var exn = Js_exn.internalToOCamlException(raw_exn);
    if (exn[0] === Aeson_decode.DecodeError) {
      return /* Error */Block.__(1, ["decodeSumWithRecordA1: " + exn[1]]);
    } else {
      throw exn;
    }
  }
  if (exit === 1) {
    return /* Ok */Block.__(0, [v]);
  }
  
}

function decodeSumWithRecordB2(json) {
  var exit = 0;
  var v;
  try {
    v = /* record */[
      /* b2 */Aeson_decode.field("b2", Aeson_decode.string, json),
      /* b3 */Aeson_decode.field("b3", Aeson_decode.$$int, json)
    ];
    exit = 1;
  }
  catch (raw_exn){
    var exn = Js_exn.internalToOCamlException(raw_exn);
    if (exn[0] === Aeson_decode.DecodeError) {
      return /* Error */Block.__(1, ["decodeSumWithRecordB2: " + exn[1]]);
    } else {
      throw exn;
    }
  }
  if (exit === 1) {
    return /* Ok */Block.__(0, [v]);
  }
  
}

function decodeSumWithRecord(json) {
  var exit = 0;
  var err;
  try {
    err = Aeson_decode.field("tag", Aeson_decode.string, json);
    exit = 1;
  }
  catch (raw_exn){
    var exn = Js_exn.internalToOCamlException(raw_exn);
    if (exn[0] === Aeson_decode.DecodeError) {
      return /* Error */Block.__(1, [exn[1]]);
    } else {
      throw exn;
    }
  }
  if (exit === 1) {
    switch (err) {
      case "A1" : 
          var match = decodeSumWithRecordA1(json);
          if (match.tag) {
            return /* Error */Block.__(1, ["decodeSumWithRecord: " + match[0]]);
          } else {
            return /* Ok */Block.__(0, [/* A1 */Block.__(0, [match[0]])]);
          }
          break;
      case "B2" : 
          var match$1 = decodeSumWithRecordB2(json);
          if (match$1.tag) {
            return /* Error */Block.__(1, ["decodeSumWithRecord: " + match$1[0]]);
          } else {
            return /* Ok */Block.__(0, [/* B2 */Block.__(1, [match$1[0]])]);
          }
          break;
      default:
        return /* Error */Block.__(1, ["Unknown tag value found '" + (err + "'.")]);
    }
  }
  
}

exports.encodeSumWithRecordA1 = encodeSumWithRecordA1;
exports.encodeSumWithRecordB2 = encodeSumWithRecordB2;
exports.encodeSumWithRecord   = encodeSumWithRecord;
exports.decodeSumWithRecordA1 = decodeSumWithRecordA1;
exports.decodeSumWithRecordB2 = decodeSumWithRecordB2;
exports.decodeSumWithRecord   = decodeSumWithRecord;
/* Aeson_encode Not a pure module */
